sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadataSuite.scala

```diff
@@ -0,0 +1,158 @@
+
+    val checkpointDir = Utils.createTempDir()
+    testStream(joined)(
+      StartStream(checkpointLocation = checkpointDir.getCanonicalPath),
+      AddData(input1, 1),
+      CheckAnswer(),
+      AddData(input2, 1, 10),       // 1 arrived on input1 first, then input2, should join
+      CheckNewAnswer((1, 2, 3)),
+      StopStream
+    )
+
+    val statePath = new Path(checkpointDir.toString, "state/0")
+    val operatorMetadata = new OperatorStateMetadataReader(statePath, hadoopConf).read()
+      .asInstanceOf[OperatorStateMetadataV1]
+
+    val expectedStateStoreInfo = Array(
+      StateStoreMetadataV1("left-keyToNumValues", 0, numShufflePartitions),
+      StateStoreMetadataV1("left-keyWithIndexToValue", 0, numShufflePartitions),
+      StateStoreMetadataV1("right-keyToNumValues", 0, numShufflePartitions),
+      StateStoreMetadataV1("right-keyWithIndexToValue", 0, numShufflePartitions))
+
+    val expectedMetadata = OperatorStateMetadataV1(
+      OperatorInfoV1(0, "symmetricHashJoin"), expectedStateStoreInfo)
+    assert(sameOperatorStateMetadata(operatorMetadata, expectedMetadata))
+  }
+
+  test("Stateful operator metadata for streaming session window") {
+    val input = MemoryStream[(String, Long)]
+    val sessionWindow: Column = session_window($"eventTime", "10 seconds")
+
+    val checkpointDir = Utils.createTempDir()
+
+    val events = input.toDF()
+      .select($"_1".as("value"), $"_2".as("timestamp"))
+      .withColumn("eventTime", $"timestamp".cast("timestamp"))
+      .withWatermark("eventTime", "30 seconds")
+      .selectExpr("explode(split(value, ' ')) AS sessionId", "eventTime")
+
+    val streamingDf = events
+      .groupBy(sessionWindow as Symbol("session"), $"sessionId")
+      .agg(count("*").as("numEvents"))
+      .selectExpr("sessionId", "CAST(session.start AS LONG)", "CAST(session.end AS LONG)",
+        "CAST(session.end AS LONG) - CAST(session.start AS LONG) AS durationMs",
+        "numEvents")
+
+    testStream(streamingDf, OutputMode.Complete())(
+      StartStream(checkpointLocation = checkpointDir.toString),
+      AddData(input,
+        ("hello world spark streaming", 40L),
+        ("world hello structured streaming", 41L)
+      ),
+      CheckNewAnswer(
+        ("hello", 40, 51, 11, 2),
+        ("world", 40, 51, 11, 2),
+        ("streaming", 40, 51, 11, 2),
+        ("spark", 40, 50, 10, 1),
+        ("structured", 41, 51, 10, 1)
+      ),
+      StopStream
+    )
+
+    val statePath = new Path(checkpointDir.toString, "state/0")
+    val operatorMetadata = new OperatorStateMetadataReader(statePath, hadoopConf).read()
+      .asInstanceOf[OperatorStateMetadataV1]
+
+    val expectedMetadata = OperatorStateMetadataV1(
+      OperatorInfoV1(0, "sessionWindowStateStoreSaveExec"),
+      Array(StateStoreMetadataV1("default", 1, spark.sessionState.conf.numShufflePartitions))
+    )
+    assert(sameOperatorStateMetadata(operatorMetadata, expectedMetadata))
+  }
```

Comment:
Let's add another test to verify multiple stateful operators (opId being 0, 1, 2, etc.).

Reply:
Multiple operators test added.